execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,564 @@
+            }
+            tmp
+        };
+
+        scratch
+            .commit(correlation_id, root_hash, effects.clone())
+            .unwrap();
+
+        let updated_hash = state.commit(correlation_id, root_hash, effects).unwrap();
+        let updated_checkout = state.checkout(updated_hash).unwrap().unwrap();
+        let all_keys = updated_checkout
+            .keys_with_prefix(correlation_id, &[])
+            .unwrap();
+
+        let stored_values = scratch.into_inner();
+        assert_eq!(all_keys.len(), stored_values.len());
+
+        for key in all_keys {
+            assert!(stored_values.get(&key).is_some());
+            assert_eq!(
+                stored_values.get(&key),
+                updated_checkout
+                    .read(correlation_id, &key)
+                    .unwrap()
+                    .as_ref()
+            );
+        }
+
+        for TestPair { key, value } in test_pairs_updated.iter().cloned() {
+            assert_eq!(
+                Some(value),
+                updated_checkout.read(correlation_id, &key).unwrap()
+            );
+        }
+    }
+
+    #[test]
+    fn commit_updates_state_with_add() {
+        let correlation_id = CorrelationId::new();
+        let test_pairs_updated = create_test_pairs_updated();
+
+        // create two lmdb instances, with a scratch instance on the first
+        let TestState { state, root_hash } = create_test_state();
+        let TestState {
+            state: state2,
+            root_hash: state_2_root_hash,
+        } = create_test_state();
+
+        let scratch = state.create_scratch();
+
+        let effects: AdditiveMap<Key, Transform> = {
+            let mut tmp = AdditiveMap::new();
+            for TestPair { key, value } in &test_pairs_updated {
+                tmp.insert(*key, Transform::Write(value.to_owned()));
+            }
+            tmp
+        };
+
+        // Commit effects to both databases.
+        scratch
+            .commit(correlation_id, root_hash, effects.clone())
+            .unwrap();
+        let updated_hash = state2
+            .commit(correlation_id, state_2_root_hash, effects)
+            .unwrap();
+
+        // Create add transforms as well
+        let add_effects: AdditiveMap<Key, Transform> = {
+            let mut tmp = AdditiveMap::new();
+            for TestPair { key, value } in &test_pairs_updated {
+                if let StoredValue::CLValue(cl_value) = value {
+                    match cl_value.cl_type() {
+                        CLType::I32 => {
+                            tmp.insert(*key, Transform::AddInt32(2i32));
+                        }
+                        CLType::String => {}
+                        _ => unreachable!(
+                            "remember to update the test pairs if new cl_types are added"
+                        ),
+                    }
+                }
+            }
+            tmp
+        };
```

Comment:
This might be much simpler and cleaner if instead of having `CLValue`s in `test_pairs_updated` we used `Transform`s.